### PR TITLE
Mirror CircleCI workflow as a GHActions workflow

### PR DESCRIPTION
**Pull request by V-FEXrt**
This PR moves all of our CircleCI jobs over to GH Actions. It doesn't delete any of the CircleCI jobs for now. I want to punt that until later in case something goes wrong and we need to revert. So for an amount of time, we will have CI jobs in both.

I have turned off the doc deploy step in CircleCI however as both GHActions and CircleCI would be fighting over the single resource.

Main changes
 - All builds were ported to GH Actions 
 - Doc deploy was turned off in CircleCI and recreated in GHActions
 - A *new* job was created that will automatically build and create a draft release whenever a tag is created. This should make it easier to create new wake releases.

### CODE REVIEW

**Comment by JakeSiFive** [comment]
I'll prioritize reviewing this today

File: .github/workflows/build.yaml
Comment (richardxia) [inline .github/workflows/build.yaml:230]: Just trying to think of how to DRY things up, and for this step, why do we need to do this after extracting the artifact? I don't know how GitHub's artifacts work under the hood, but normally archive file formats (e.g. tar) should preserve the executability bit.

**Review by richardxia — APPROVED** [review]
I didn't review this in great detail, since I assume that the tricky bits with the exact `docker` commands were copies verbatim from the CircleCI jobs, but this overall LGTM!

Upon first glance, it feels like there's a lot of repeated boilerplate, but I don't think there's really a good way to avoid it because there are just enough little differences between the different distros that would defeat any benefit from refactoring.

One last nit-picky comment: The title of the PR is misleading, since "Actions" and "Workflows" are different kinds of things in GH Actions. What you've changed here is a Workflow, and a Custom Action would be more like a reusable step that you call from the workflow.

**Comment by V-FEXrt** [comment]
>  The title of the PR is misleading, since "Actions" and "Workflows" are different kinds of things in GH Actions

yeah, my bad. This PR started out very different and I forgot to update the title

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:230]: > I don't know how GitHub's artifacts work under the hood, but normally archive file formats (e.g. tar) should preserve the executability bit.

It gets lost in in the upload flow :(, its a common enough question that it has a top level note in the documentation. https://github.com/actions/upload-artifact#permission-loss

Comment (richardxia) [reply .github/workflows/build.yaml:230]: I had suspected and was afraid you were going to say that :(

Comment (V-FEXrt) [reply .github/workflows/build.yaml:230]: I do want to DRY here if possible.

The best option is via a `matrix` which basically repeats a job but with different parameters. And its _almost_ here with the following params.

- docker image
 - build command
 - test command
 - artificats glob
 - install destination
 - upload destination
 
 We could certainly just do extra steps like adding the tests/debian dirs even when they aren't needed to cover the differences, but with 6 parameters its hard for me to decided if its worth it. I'm personally leaning towards yes

Comment (richardxia) [reply .github/workflows/build.yaml:230]: If you're willing to spend time to DRY it up, I'd support that effort, since with the 7 or so different environments, I think it'd be worth it. But if it ends up being too difficult, I don't think it would be the worst thing to leave things as they are right now.

Comment (V-FEXrt) [reply .github/workflows/build.yaml]: This should be done now!

File: .circleci/config.yml
Comment (ag-eitilt) [inline .circleci/config.yml:254]: If we weren't about to delete this file, I'd be more concerned about the blank line at the end. 😉

File: .github/workflows/build.yaml
Comment (ag-eitilt) [inline .github/workflows/build.yaml:266]: I assume this catches PRs being merged as well?

File: .github/workflows/build.yaml
Comment (ag-eitilt) [inline .github/workflows/build.yaml:286]: I foresee a need to update the "Make a release" workflow docs.  That's probably a good thing; this *feels* like it should allow a bit easier releases.

File: .github/workflows/build.yaml
Comment (ag-eitilt) [inline .github/workflows/build.yaml:308]: I don't see any typos beyond this one.
```suggestion
            path: alpine
```

File: .github/workflows/build.yaml
Comment (ag-eitilt) [inline .github/workflows/build.yaml:90]: The syntax highlighter is complaining about these being empty; I imagine the keys are in here because you want them here, but would explicit empty strings break anything?  (It fixes the highlighter, but I'm not sure about the Workflows parser.)
```suggestion
              extra_docker_build_args: ""
              extra_docker_run_args: ""
```

**Review by ag-eitilt — COMMENTED** [review]
Don't want to block this in case I don't see the fix for a bit, but there are a couple, very small fixes which you might want to make.  On the whole, though, this looks promising!  (I've not seen what it looks like in the CI interface yet.)

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:266]: Right so the whole workflow triggers on any of 

1) PR opened
2) commit pushed to master
3) Tag starting with a `v` created and pushed

The filter for this specific job says only run when a commit is pushed to master, and since we only let commits land on master via a PR this runs right after a PR is merged

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:286]: Yep, I'm planning on making the update after this lands. I need to land this before update the docs to make sure I get the details right since I can only test the full flow after it lands,

**Comment by V-FEXrt** [comment]
I'm going to remove `MacOS build` as required since it was renamed to `MacOS 10.15 Build`. Once this lands on master, I'll re-add it as required with the correct name. That was I can land this without blocking other PRs in the meantime

File: .github/workflows/build.yaml
Comment (JakeSiFive) [inline .github/workflows/build.yaml:50]: What are retention-days?

File: .github/workflows/build.yaml
Comment (JakeSiFive) [inline .github/workflows/build.yaml:40]: What does the upload-artifact action do?

File: .github/workflows/build.yaml
Comment (JakeSiFive) [inline .github/workflows/build.yaml:293]: what does download-artifact do?

File: .github/workflows/build.yaml
Comment (JakeSiFive) [reply .github/workflows/build.yaml:354]: MIT license and yeah this is someone else's action. I reviewed the action code and it looks fine and seems to have a lot of users but I think we should consider keeping our own fork of this

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:50]: number of days the artifact is kept before being deleted. Default (and max) is 90. I set it to 1 for artifacts that are only used as intermediates for other steps to be a good steward

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:40]: uploads the contents of a folder or path to a key-value store that is a new store for each workflow run

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:162]: yeah, old commit where each of this were one a separate job

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:220]: Unclear. This is just the rule we had before. I don't really want to hit on that with this PR and we should review this separately

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml:293]: reverse of upload-artifact

File: .github/workflows/build.yaml
Comment (JakeSiFive) [reply .github/workflows/build.yaml:220]: Sure sounds good. Worst case can't be any worse than what we had before

File: .github/workflows/build.yaml
Comment (JakeSiFive) [reply .github/workflows/build.yaml:354]: Ashley and I spoke offline about this. I'm still in favor of forking but if enough others or if Matthew feels that a commit hash is an acceptable means of pinning the action code then I'm down. If we're not down to fork or if we can't reach consensus on this soon enough we'll just remove this code and fall back to manually uploading like we were before. This would be a nice plus though.

File: .github/workflows/build.yaml
Comment (ag-eitilt) [reply .github/workflows/build.yaml]: > > I don't know how GitHub's artifacts work under the hood, but normally archive file formats (e.g. tar) should preserve the executability bit.
> 
> It gets lost in in the upload flow :(, its a common enough question that it has a top level note in the documentation. https://github.com/actions/upload-artifact#permission-loss

[A few sections down](https://github.com/actions/upload-artifact#maintaining-file-permissions-and-case-sensitive-files), it looks like that might only apply to the raw uploads, and making a tarball is actually given as a solution for maintaining the permissions.  I'm sure you've included this line as a reaction to trying and failing to run the tests without it, so there's probably something else going on which is *also* stripping those permissions before the files get packaged up, but that just underscores how complex this whole remote-execution-via-ostensibly-data-files system is.  (Very handy though, no matter its foibles!)

File: .github/workflows/build.yaml
Comment (JakeSiFive) [reply .github/workflows/build.yaml:354]: Further research reveals that git is using SHA-1DC which is somehow hardened against the SHA1 attacks but is otherwise compatible and at present has no publicly known attacks against it. So a commit hash is probably secure enough for our purposes.

**Review by JakeSiFive — APPROVED** [review]
I don't know what the latest error is about but this all looks good to me now once its working

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml]: Done! PTAL!

Comment (V-FEXrt) [reply .github/workflows/build.yaml]: nvm, just saw the most recent comment. Thanks!

File: .github/workflows/build.yaml
Comment (V-FEXrt) [reply .github/workflows/build.yaml]: yeah adding it was in response to a failed test. We can explore improving it later, but it seems like a small issue and I'd like to get this landed to minimize the time with the CI in a weird state.

**Pull request merged**
